File: src/algorithms/tree.js
const Tree = require('../models/tree')

const computeConcatenatedWords = (words) => {
  const tree = new Tree()

  tree.addWords(words)

  let concatenatedWordCount = 0
  let maxConcatLength = 0, secondMaxConcatLength = 0
  let maxLengthConcatWords = [], secondMaxLengthConcatWords = []

  for (let i = 0; i < words.length; i++) {
    const word = words[i]
    if (tree.isConcatenated(word)) {
      concatenatedWordCount++
      if (word.length > maxConcatLength) {
        secondMaxConcatLength = maxConcatLength
        secondMaxLengthConcatWords = maxLengthConcatWords
        maxConcatLength = word.length
        maxLengthConcatWords = [word]
      } else if (word.length === maxConcatLength) {
        maxLengthConcatWords.push(word)
      } else if (word.length > secondMaxConcatLength) {
        secondMaxConcatLength = word.length
        secondMaxLengthConcatWords = [word]
      } else if (word.length === secondMaxConcatLength) {
        secondMaxLengthConcatWords.push(word)
      }
    }
  }

  return {
    concatenatedWordCount,
    maxLengthConcatWords,
    secondMaxLengthConcatWords
  }
}

module.exports = {
  computeConcatenatedWords
}

File: py_src/algorithms/tree.py
from classes.tree import Tree

def compute_concatenated_words(words):

    tree = Tree()
    tree.add_words(words)

    concatenated_word_count = 0
    max_concat_length = 0
    second_max_concat_length = 0
    max_length_concat_words = []
    second_max_length_concat_words = []

    for word in words:

        if tree.is_concatenated(word):
            concatenated_word_count += 1

            if len(word) > max_concat_length:
                second_max_concat_length = max_concat_length
                second_max_length_concat_words = max_length_concat_words
                max_concat_length = len(word)
                max_length_concat_words = [word]

            elif len(word) == max_concat_length:
                max_length_concat_words.append(word)

            elif len(word) > second_max_concat_length:
                second_max_concat_length = len(word)
                second_max_length_concat_words = [word]

            elif len(word) == second_max_concat_length:
                second_max_length_concat_words.append(word)

    return concatenated_word_count, \
        max_length_concat_words, \
        second_max_length_concat_words


